feat(curriculum): add magazine layout lab (#57174)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-magazine-layout/6735a7370e0ae93a4577c689.md

**User Stories:**

1. You should have a `main` element with the class `magazine-cover`.

2. Your `.magazine-cover` should contain a `header` element for the title and `section` elements for each article and the cover image.

3. You should set the display property of the `.magazine-cover` to `grid.`

4. The `.magazine-cover` should define a grid layout with the following structure:

   - The title spanning the top row.
   - Two feature articles and a cover image in the second row.
   - Two small articles and the cover image in the third row.

5. The `.magazine-cover` should have a `grid-template-columns` property that divides the space into three equal parts.

6. The `.magazine-cover` should have a `grid-template-rows` property. It should be set to `auto` for the first row, and divide equally the space occupied by the remaining two rows.

7. You should add a gap between grid items and set the width of the `.magazine-cover` to occupy most of the viewport width.

**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.

# --hints--

You should have a `main` element.

```js
assert.exists(document.querySelector('main'));
```

> You should have a `.magazine-cover` class.

Reviewer: This is not checking that the main element has the requested class

Author: that is being checked in the next step

```js
assert.exists(document.querySelector('.magazine-cover'));
```

You should have a `header` element for the title.

```js

> assert.exists(document.querySelector('header'));

Reviewer: This should check that the header is a child of the main element

```

> You should have a `section` element for each article and the cover image, with classes `feature-article`, `small-article1`, `small-article2`, and `cover-image` respectively.

Reviewer: Users stories don't specify exactly what you need to do with articles and classes. It should be clarified from the beginning.

Author: The user stories only highlight css grid related sections, but I can re-edit them to cover all section if it's needed.

Reviewer 2: Yes please, add everything that the camper need to do to the user stories, it's really confusing to test for something that is not in the user stories

```js
const featureArticle = document.querySelector('section.feature-article');
const smallArticle1 = document.querySelector('section.small-article1');
const smallArticle2 = document.querySelector('section.small-article2');

> const coverImage = document.querySelector('section.cover-image');

Reviewer: missing asserts in this  test

---
id: 6735a7370e0ae93a4577c689

> title: Design a Magazine Layout Using CSS Grid

Reviewer 2: ```suggestion
title: Design a Magazine Layout
```

challengeType: 14

> dashedName: design-a-magazine-layout-using-css-grid

Reviewer 2: ```suggestion
dashedName: design-a-magazine-layout
```

3. You should set the display property of the `.magazine-cover` to `grid.`

4. The `.magazine-cover` should define a grid layout with the following structure:

   - The title spanning the top row.
   - Two feature articles and a cover image in the second row.
   - Two small articles and the cover image in the third row.

5. The `.magazine-cover` should have a `grid-template-columns` property that divides the space into three equal parts.

6. The `.magazine-cover` should have a `grid-template-rows` property. It should be set to `auto` for the first row, and divide equally the space occupied by the remaining two rows.

7. You should add a gap between grid items and set the width of the `.magazine-cover` to occupy most of the viewport width.

**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.

# --hints--

You should have a `main` element.

```js
assert.exists(document.querySelector('main'));
```

You should have a `.magazine-cover` class for the main element.

```js
assert.exists(document.querySelector('.magazine-cover'));
```

You should have a `header` element for the title.

```js

> assert.exists(new __helpers.CSSHelp(document).getStyle('main > header'));

Reviewer: ```suggestion
assert.exists(document.querySelector("main > header"));
```

> new __helpers.CSSHelp(document).exists('main > header'),

Reviewer 2: missing assert
```suggestion
assert.exists(new __helpers.CSSHelp(document).getStyle('main > header'));
```

Reviewer 2: if you fix this the CI tests should pass

```

You should have a `section` element for each article and the cover image, with classes `feature-article`, `small-article1`, `small-article2`, and `cover-image` respectively.

```js
const featureArticle = document.querySelector('section.feature-article');
const smallArticle1 = document.querySelector('section.small-article1');
const smallArticle2 = document.querySelector('section.small-article2');
const coverImage = document.querySelector('section.cover-image');

assert.exists(featureArticle);
assert.exists(smallArticle1);
assert.exists(smallArticle2);
assert.exists(coverImage);
```

The display property of the `.magazine-cover` should be set to `grid`.

```js
assert.equal(getComputedStyle(document.querySelector('.magazine-cover')).display, 'grid');
```

Ensure that you have a `grid-template-areas` property with `title` spanning all three columns.

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.equal(magazineCoverStyle.gridTemplateAreas.includes('"title title title"'), true);

Reviewer 2: ```suggestion
assert.include(magazineCoverStyle.gridTemplateAreas, '"title title title"');
```

```

The second row of the grid template should contain two `feature article` and a `cover image`.1

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.equal(magazineCoverStyle.gridTemplateAreas.includes('"feature-article feature-article cover-image"'), true);

Reviewer 2: ```suggestion
assert.include(magazineCoverStyle.gridTemplateAreas, '"feature-article feature-article cover-image"');
```

```

The third row of the grid template should contain `small-article1`, `small-article2`, and `cover-image`.

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.equal(magazineCoverStyle.gridTemplateAreas.includes('"small-article1 small-article2 cover-image"'), true);

Reviewer 2: ```suggestion
assert.include(magazineCoverStyle.gridTemplateAreas, '"small-article1 small-article2 cover-image"');
```

```

Your `.magazine-cover` class should have a `grid-template-columns` property set to `1fr 1fr 1fr`.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.magazine-cover')?.gridTemplateColumns === '1fr 1fr 1fr');

Reviewer 2: ```suggestion
assert.equal(new __helpers.CSSHelp(document).getStyle('.magazine-cover')?.gridTemplateColumns, '1fr 1fr 1fr');
```

```

Your `.magazine-cover` class should have a `grid-template-rows` property set to `auto 1fr 1fr`.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.magazine-cover')?.gridTemplateRows === 'auto 1fr 1fr');

Reviewer 2: ```suggestion
assert.equal(new __helpers.CSSHelp(document).getStyle('.magazine-cover')?.gridTemplateRows, 'auto 1fr 1fr');
```

```

You should add a gap of `10px` between grid items.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.magazine-cover')?.gap === '10px');

Reviewer 2: ```suggestion
assert.equal(new __helpers.CSSHelp(document).getStyle('.magazine-cover')?.gap, '10px');
```

---
id: 6735a7370e0ae93a4577c689
title: Design a Magazine Layout
challengeType: 25
dashedName: design-a-magazine-layout
demoType: onClick
---

# --description--

In this lab, you will design a magazine layout using CSS Grid, including concepts like grid rows and grid columns.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should have a `main` element with the class `magazine-cover`.

2. Your `.magazine-cover` should contain a `header` element for the title with the class `title`.

> 3. You should have `section` elements for each article and the cover image. Each section should contain a class attribute for each article and the cover image: `feature-article`, `small-article1`, `small-article2`, and `cover-image` respectively.

Reviewer 3: ```suggestion
3. You should have `section` elements for each article and the cover image. Each individual section should contain the following class attributes for each article and the cover image: `feature-article`, `small-article1`, `small-article2`, and `cover-image` respectively.
```

**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.

# --hints--

You should have a `main` element.

```js
assert.exists(document.querySelector('main'));
```

You should have a `.magazine-cover` class for the main element.

```js
assert.exists(document.querySelector('.magazine-cover'));
```

You should have a `header` element for the title.

```js
assert.exists(document.querySelector("main > header"));
```

You should have a `section` element for each article and the cover image, with classes `feature-article`, `small-article1`, `small-article2`, and `cover-image` respectively.

```js
const featureArticle = document.querySelector('section.feature-article');
const smallArticle1 = document.querySelector('section.small-article1');
const smallArticle2 = document.querySelector('section.small-article2');
const coverImage = document.querySelector('section.cover-image');

assert.exists(featureArticle);
assert.exists(smallArticle1);
assert.exists(smallArticle2);
assert.exists(coverImage);
```

The display property of the `.magazine-cover` should be set to `grid`.

```js
assert.equal(getComputedStyle(document.querySelector('.magazine-cover')).display, 'grid');
```

> Ensure that you have a `grid-template-areas` property with `title` spanning all three columns.

Reviewer 3: It should be mentioned in the instructions that you should use the `grid-template-areas` property.

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.include(magazineCoverStyle.gridTemplateAreas, '"title title title"');

Reviewer: ```suggestion
assert.match(magazineCoverStyle.gridTemplateAreas, /^"title title title"/);
```
Otherwise it doesn't check that is the correct row. Same thing for the other rows.

```

> The second row of the grid template should contain two `feature article` and a `cover image`.1

Reviewer 3: ```suggestion
The second row of the grid template should contain two `feature article` and a `cover image`
```

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));
assert.include(magazineCoverStyle.gridTemplateAreas, '"feature-article feature-article cover-image"');
```

The third row of the grid template should contain `small-article1`, `small-article2`, and `cover-image`.

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.include(magazineCoverStyle.gridTemplateAreas, '"small-article1 small-article2 cover-image"');

Reviewer: ```suggestion
assert.match(magazineCoverStyle.gridTemplateAreas, /"small-article1 small-article2 cover-image"$/);
```

---
id: 6735a7370e0ae93a4577c689
title: Design a Magazine Layout
challengeType: 25
dashedName: design-a-magazine-layout
demoType: onClick
---

# --description--

In this lab, you will design a magazine layout using CSS Grid, including concepts like grid rows and grid columns.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should have a `main` element with the class `magazine-cover`.

2. Your `.magazine-cover` should contain a `header` element for the title with the class `title`.

3. You should have four `section` elements in your page. The first three sections for magazine articles and a fourth `section` element for a cover image.  

> 4. The four  `section` elements should have a `class` attribute of `feature-article`, `small-article1`, `small-article2`, and `cover-image`, respectively.

Reviewer: ```suggestion
4. The four `section` elements should have a `class` attribute of `feature-article`, `small-article1`, `small-article2`, and `cover-image`, respectively.
```

> 5. You should set the display property of the `.magazine-cover` to `grid.`

Reviewer: ```suggestion
5. You should set the `display` property of `.magazine-cover` to `grid.`
```

> 6. The `.magazine-cover` should have a `grid-template-areas` property that defines the layout of the grid:

Reviewer 3: From this instruction, I understood that there should be 2 featured posts in the second row. However, the solution/ demo asks for only one.

This is the layout I came up with and it doesn't pass:

<details>

<summary>Image</summary>

![image](https://github.com/user-attachments/assets/de0e3acc-6b8f-43c3-b0d5-be3ff3618f9c)

</details>

Reviewer: Oh you're right. Maybe it should say `A feature article spanning two columns and a cover image in the second row`

Reviewer 3: Yes, that would make the instruction clear.

Reviewer: I left two suggestions. If those are good I think we can go ahead so we merge this PR @zairahira

   - The title spanning the top row.

>    - Two feature articles and a cover image in the second row.

Reviewer: ```suggestion
   - A feature article spanning two columns and a cover image in the second row.
```

# --hints--

You should have a `main` element.

```js
assert.exists(document.querySelector('main'));
```

Your `main` element should have a `class` of `magazine-cover`.

```js
assert.exists(document.querySelector('main.magazine-cover'));
```

You should have a `header` element within your `main` element.

```js
assert.exists(document.querySelector("main > header"));
```

Your `header` element should have a `class` of `title` and should contain some text.

```js
const header = document.querySelector("header.title");
assert.exists(header);
assert.isAbove(header.innerText.length, 0);
```

You should have a `section` element for each article and the cover image, with classes `feature-article`, `small-article1`, `small-article2`, and `cover-image` respectively.

```js
const featureArticle = document.querySelector('section.feature-article');
const smallArticle1 = document.querySelector('section.small-article1');
const smallArticle2 = document.querySelector('section.small-article2');
const coverImage = document.querySelector('section.cover-image');

assert.exists(featureArticle);
assert.exists(smallArticle1);
assert.exists(smallArticle2);
assert.exists(coverImage);
```

> The display property of the `.magazine-cover` should be set to `grid`.

Reviewer: ```suggestion
The `display` property of the `.magazine-cover` should be set to `grid`.
```

```js
assert.equal(getComputedStyle(document.querySelector('.magazine-cover')).display, 'grid');
```

Your `.magazine-cover` should have a `grid-template-areas` property with `title` spanning all three columns of the first row.

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.match(magazineCoverStyle.gridTemplateAreas, /^"title title title"/);

Reviewer: ```suggestion
assert.match(magazineCoverStyle.gridTemplateAreas, /^("|')title title title\1/);
```
I forgot to include single quotes. Same for the other regular expressions

```

> The second row of the grid template should contain two `feature article` and a `cover image`.

Reviewer: ```suggestion
The second row of the grid template should contain a `feature article` spanning two columns and a `cover image`.
```

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.match(magazineCoverStyle.gridTemplateAreas, /."feature-article feature-article cover-image"./);

Reviewer: ```suggestion
assert.match(magazineCoverStyle.gridTemplateAreas, /.("|')feature-article feature-article cover-image\1./);
```

```

The third row of the grid template should contain `small-article1`, `small-article2`, and `cover-image`.

```js
const magazineCoverStyle = getComputedStyle(document.querySelector('.magazine-cover'));

> assert.match(magazineCoverStyle.gridTemplateAreas, /"small-article1 small-article2 cover-image"$/);

Reviewer: ```suggestion
assert.match(magazineCoverStyle.gridTemplateAreas, /("|')small-article1 small-article2 cover-image\1$/);
```
